fix == problem created by FORMULATEXT

RiverCrimea opened the pull request:
If the text to be converted to a formula was created by the FORMULATEXT function, it already starts with a =. Since textToFunction adds another = to the beginning of everything it converts, the formula will begin with == and create an error instead of a working formula.
Added a line to check for the presence of == after every conversion and replace it with =, if found.

RiverCrimea committed:
fix == problem created by FORMULATEXT

If the text to be converted to a formula was created by the FORMULATEXT function, it already starts with a =. Since textToFunction adds another = to the beginning of everything it converts, the formula will begin with == and create an error instead of a working formula.
Added a line to check for the presence of == after every conversion and replace it with =, if found.

RiverCrimea commented:
fixes Will-Renius/TextFormulaConverter#3

Will-Renius reviewed (commented):
Are there other situations "==" could occur? perhaps we should only check if first two characters are "=="

RiverCrimea commented:
Oh, I see what you're saying.
Yes, it is possible that a formula contains a string with "==" in it. Checking only the first 2 characters is going to be much more complicated, though. Let me think for a moment.

RiverCrimea committed:
change FORMULATEXT fix

Instead of automatically replacing all occurrences of == with = in a formula, which could catch intended string values, check only the first two characters for ==. If the condition exists, then remove one character from the beginning.

RiverCrimea commented:
Okay, how's that? I'm not completely familiar with the script syntax, and I'm not in a position to look it up right now, so I might have got the if statement wrong. Apart from possibly adjusting the if statement though, I think that will work, even if it is a bit clunky.

RiverCrimea committed:
fix syntax error

Apparently, I write code in my sleep now. Sorry about that.

RiverCrimea commented:
I attempted multiple times, in multiple ways, to create a link between the issue and the PR, but GitHub still doesn't show that the link exists. Maybe it's because I lack proper permissions. Maybe it's because I'm a fucking idiot who can't figure out what's right in front of his face. In any case, I've assembled the best fix I know how to make. If it works, then great! If it doesn't, then I'm sorry I wasted your time.

RiverCrimea committed:
update to correct code

Previously used code syntax borrowed from somewhere in my memory instead of the actual code syntax required.
Code updated. Script tested. Script verified to be working with updated code.
Apologies for my n00biness.

Will-Renius commented:
I see the pull request now. I don't have any automated tests for this so I'll run some manual tests and try to get the package updated.

RiverCrimea commented:
If it helps you, I fully tested the code with my changes before my most recent commit.

Will-Renius commented:
Can you share a sheet that I can use to replicate the error? It's supposed to recognize that it's a formula and not try to convert.

RiverCrimea commented:
[Here is a test document I created.](https://docs.google.com/spreadsheets/d/1mH5YnEvKs46U-jADCx8QGp2bH9RORWQs88_7gqit878/edit?usp=sharing) I installed the existing TextFormulaConverter add-on. Then I used the script editor to create "tlc-fix". You will need to authorize the script to run on your account, and then the evidence that it works should be clear. ;-)

RiverCrimea committed:
create .gitignore

RiverCrimea committed:
update .gitignore

RiverCrimea commented:
I looked at the sharing options for my test document again. Short version: I'm pretty sure they were wrong.
They are correct now.

Will-Renius commented:
I have access now, but I still don't see how you're getting it to add an
extra equal sign. Converting between formula and text works fine for me on
your spreadsheet

On Sat, Aug 22, 2020 at 5:06 PM PellaAndroid <notifications@github.com>
wrote:

> I looked at the sharing options for my test document again. Short version:
> I'm pretty sure they were wrong.
> They are correct now.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/Will-Renius/TextFormulaConverter/pull/4*issuecomment-678694635__;Iw!!HXCxUKc!g2U92KmOXQ5pFXBOeh5jIxuQOkRXblBWOtXMoypuwNf3mtlqbxKuLOiUfRzqc94$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ACVMYHL76N5PHSCCZ5QFFL3SCAXNTANCNFSM4NN35HEQ__;!!HXCxUKc!g2U92KmOXQ5pFXBOeh5jIxuQOkRXblBWOtXMoypuwNf3mtlqbxKuLOiUu-HBbrs$>
> .
>


-- 
-Will Renius

RiverCrimea commented:
All right, I finally have this figured out.
I think Google made some changes that made my test document go weird, so I created a [new test document](https://docs.google.com/spreadsheets/d/18YCaPIeQ2fz_NoenjplooFtFjAGEuyVsNTNaOejRY8Q/edit?usp=sharing).

It's a very specific sequence of steps.

1. Begin with a formula. Any formula.
2. Use the FORMULATEXT function to convert the formula to text.
3. We cannot use what step 2 created to create a new formula because Sheets will give us the FORMULATEXT formula. Therefore, copy the result of step 2 and paste _**values only**_.
4. Use TFC "Convert to formula" on the result of step 3. TFC adds `=` to the beginning, even though Sheets already has one there. This creates `#ERROR!`.
 
Open the Script Editor to find function `tfc-fix`. Run `textToFunction` on the result of step 3. (You will need to click past the warnings to authorize the function to work on your account.) After adding the unnecessary (in this case) `=` to the beginning of the formula, `tfc-fix` checks for the presence of `==` at the beginning and strips one of them.

My deepest apologies that this has taken so long, but there it is.

RiverCrimea commented:
It's been a while. LOL
Have you had a chance to run the quick test I described, sir??
